Release CI

hasufell opened the pull request.

hasufell committed:
MAKE: Fix header copying

hasufell committed:
TESTS: Skip uniques test if no git repo

hasufell committed:
TESTS: Makefile fixes

hasufell committed:
ghc-toolchain: Drop `ld.gold` from merge object command

It's deprecated.

hasufell committed:
CI: Add release workflow

hasufell committed:
Fix FreeBSD stage2/stage3

hasufell committed:
rts: check if we need -lm in configure

hasufell committed:
TESTS: Add missing EXTRA_HC_OPTS

angerman commented on utils/ghc-toolchain/src/GHC/Toolchain/Tools/MergeObjs.hs:
```diff
@@ -22,7 +22,7 @@ data MergeObjs = MergeObjs { mergeObjsProgram :: Program
 findMergeObjs :: ProgOpt -> Cc -> CcLink -> Nm -> M MergeObjs
 findMergeObjs progOpt cc ccLink nm = checking "for linker for merging objects" $ do
-    prog <- findProgram "linker for merging objects" progOpt ["ld.gold", "ld"]
+    prog <- findProgram "linker for merging objects" progOpt ["ld"]
```
This should ultimately also end up being a seperate PR for upstreaming.

hasufell replied:
https://gitlab.haskell.org/ghc/ghc/-/merge_requests/14942

angerman commented on cabal.project.stage2:
```diff
@@ -128,6 +128,22 @@ if os(darwin)
     flags: +tables-next-to-code +leading-underscore
 	ld-options: -undefined warning
 
+if os(freebsd)
+  package rts
+    ghc-options: "-optc-DProjectVersion=\"914\""
+    ghc-options: "-optc-DRtsWay=\"v\""
+    ghc-options: "-optc-DHostPlatform=\"x86_64-portbld-freebsd\""
+    ghc-options: "-optc-DHostArch=\"x86_64\""
+    ghc-options: "-optc-DHostOS=\"freebsd\""
+    ghc-options: "-optc-DHostVendor=\"unknown\""
+    ghc-options: "-optc-DBuildPlatform=\"FIXME\""
+    ghc-options: "-optc-DBuildArch=\"FIXME\""
+    ghc-options: "-optc-DBuildOS=\"FIXME\""
+    ghc-options: "-optc-DBuildVendor=\"FIXME\""
+    ghc-options: "-optc-DGhcUnregisterised=\"FIXME\""
+    ghc-options: "-optc-DTablesNextToCode=\"FIXME\""
+    ghc-options: "-optc-DFS_NAMESPACE=rts"
+    flags: +tables-next-to-code
```
These should be extracted into a separate PR: "FreeBSD support".

hasufell replied:
Don't see why.

angerman commented on Makefile:
```diff
@@ -100,6 +100,7 @@ ROOT_DIR := $(patsubst %/,%,$(dir $(realpath $(lastword $(MAKEFILE_LIST)))))
 GHC0 ?= ghc-9.8.4
 PYTHON ?= python3
 CABAL ?= cabal
+SED ?= sed
 
```
These seem sensible. We should probably extract this change? Seems already like a PR we could merge?

angerman reviewed (commented):
Looks good so far. We might want to extract parts.

hasufell committed:
rts: Fix preprocessor flags

hasufell committed:
TESTS/CI: Skip T14999

hasufell committed:
Support statically linking executables properly

Fixes #26434

In detail, this does a number of things:
* Makes GHC aware of 'extra-libraries-static' (this changes the package
  database format).
* Adds a switch '-static-external' that will honour 'extra-libraries-static'
  to link external system dependencies statically.
* Adds a new field to settings/targets: "ld supports verbatim namespace".
  This field is used by '-static-external' to conditionally use '-l:foo.a'
  syntax during linking, which is more robust than trying to find the
  absolute path to an archive on our own.
* Adds a switch '-fully-static' that is meant as a high-level interface
  for e.g. cabal. This also honours 'extra-libraries-static'.

This also attempts to clean up the confusion around library search directories.
At the moment, we have 3 types of directories in the package database
format:
* library-dirs
* library-dirs-static
* dynamic-library-dirs

However, we only have two types of linking: dynamic or static. Given the
existing logic in 'mungeDynLibFields', this patch assumes that
'library-dirs' is really just nothing but a fallback and always
prefers the more specific variants if they exist and are non-empty.

Conceptually, we should be ok with even just one search dirs variant.
Haskell libraries are named differently depending on whether they're
static or dynamic, so GHC can conveniently pick the right one depending
on the linking needs. That means we don't really need to play tricks
with search paths to convince the compiler to do linking as we want it.
For system C libraries, the convention has been anyway to place static and
dynamic libs next to each other, so we need to deal with that issue
anyway and it is outside of our control. But this is out of the scope
of this patch.

This patch is backwards compatible with cabal. Cabal should however
be patched to use the new '-fully-static' switch.

hasufell committed:
Update cabal to support GHCs new '-fully-static'

hasufell committed:
Move wasm "ways" constraints to 'makeDynFlagsConsistent'

hasufell committed:
Warn when "-dynamic" is mixed with "-staticlib"

hasufell committed:
Rename LinkBinary/BinaryLinkMode

"Executable" seems more appropriate.

hasufell commented on testsuite/mk/boilerplate.mk:
```diff
@@ -255,7 +255,7 @@ endif
 
 $(TOP)/ghc-config/ghc-config : $(TOP)/ghc-config/ghc-config.hs
-	"$(TEST_HC)" --make -o $@ $<
+	"$(TEST_HC)" $(EXTRA_HC_OPTS) --make -o $@ $<
```
https://gitlab.haskell.org/ghc/ghc/-/merge_requests/14941

hasufell committed:
Add -exclude-static-external

hasufell committed:
Make sure 'extra-libraries-static' is consistently defined

hasufell committed:
Add test for `-static-external`

hasufell committed:
Add test for `-fully-static`

hasufell committed:
Add libffi-clib

hasufell committed:
Use bundled libffi-clib

angerman commented on libraries/libffi/configure.ac:
```diff
@@ -0,0 +1,395 @@
+      [Define this if you want to enable pax emulated trampolines (experimental)])
+  fi)
+
+LT_SYS_SYMBOL_USCORE
```
I think because the macro is part of libtool.
- https://github.com/besser82/libxcrypt/issues/125
- https://github.com/libffi/libffi/issues/210

hasufell committed:
Fix docs

hasufell committed:
Better error handling when linking statically

hasufell committed:
Update libffi-clib submodule

hasufell committed:
downstream: Use bundled libffi

hasufell committed:
GHC: enable PIC on linux/freebsd x86_64 too

hasufell committed:
Mark T25240 as fragile

hasufell committed:
Update T20604 stdout

Since we now ignore loading ANY rts dependency.

hasufell committed:
Fix mostly-static test on windows

hasufell committed:
TESTS: Mark T13786 as fragile on darwin x86_64 too

hasufell committed:
TESTS: Mark T7040_ghci fragile on darwin x86_64 as well

angerman commented:
Superceeded by #113
